File: GolfCoachNow/Network/APIClient.swift
import Foundation

enum APIConfig {
    static let baseURL = "https://golfcoachnow.pythonanywhere.com"

    static func moduleURL(for module: GolfModule) -> String {
        baseURL + module.apiEndpoint
    }

    static func uploadURL(for module: GolfModule, deviceId: String) -> String {
        baseURL + "/upload?module=" + module.uploadModuleParam + "&device_id=" + deviceId
    }
}

enum APIError: LocalizedError {
    case invalidURL
    case requestFailed(statusCode: Int)
    case serverUnreachable
    case invalidResponse
    case fileReadError

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid API URL."
        case .requestFailed(let code): return "Request failed (HTTP \(code)). Try again."
        case .serverUnreachable: return "Backend not available. Check your connection."
        case .invalidResponse: return "Invalid response from server."
        case .fileReadError: return "Could not read video file."
        }
    }
}

final class APIClient {

    static let shared = APIClient()
    private let session: URLSession

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 120
        session = URLSession(configuration: config)
    }

    func uploadVideo(fileURL: URL, module: GolfModule, completion: @escaping (Result<CorrectionResponse, Error>) -> Void) {
        let deviceId = EntitlementManager.shared.deviceId
        guard let url = URL(string: APIConfig.uploadURL(for: module, deviceId: deviceId)) else {
            completion(.failure(APIError.invalidURL))
            return
        }

        guard let fileData = try? Data(contentsOf: fileURL) else {
            completion(.failure(APIError.fileReadError))
            return
        }

        let boundary = UUID().uuidString
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        let filename = fileURL.lastPathComponent
        let mime = mimeType(for: fileURL)

        body.append("--\(boundary)\r\n")
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"\(filename)\"\r\n")
        body.append("Content-Type: \(mime)\r\n\r\n")
        body.append(fileData)
        body.append("\r\n--\(boundary)--\r\n")

        request.httpBody = body

        session.dataTask(with: request) { data, response, error in
            self.handleResponse(data: data, response: response, error: error, completion: completion)
        }.resume()
    }

    func sendModuleData(_ scores: [String: Double], module: GolfModule, completion: @escaping (Result<CorrectionResponse, Error>) -> Void) {
        guard let url = URL(string: APIConfig.moduleURL(for: module)) else {
            completion(.failure(APIError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let payload: [String: Any] = [
            "data": scores,
            "device_id": EntitlementManager.shared.deviceId,
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: payload)

        session.dataTask(with: request) { data, response, error in
            self.handleResponse(data: data, response: response, error: error, completion: completion)
        }.resume()
    }

    private func handleResponse(data: Data?, response: URLResponse?, error: Error?, completion: @escaping (Result<CorrectionResponse, Error>) -> Void) {
        if let urlError = error as? URLError,
           [.notConnectedToInternet, .cannotConnectToHost, .timedOut].contains(urlError.code) {
            completion(.failure(APIError.serverUnreachable))
            return
        }

        if let error {
            completion(.failure(error))
            return
        }

        guard let http = response as? HTTPURLResponse else {
            completion(.failure(APIError.invalidResponse))
            return
        }

        guard (200...299).contains(http.statusCode), let data else {
            completion(.failure(APIError.requestFailed(statusCode: http.statusCode)))
            return
        }

        do {
            let decoder = JSONDecoder()
            decoder.keyDecodingStrategy = .convertFromSnakeCase
            let result = try decoder.decode(CorrectionResponse.self, from: data)
            completion(.success(result))
        } catch {
            completion(.failure(APIError.invalidResponse))
        }
    }

    // MARK: - Talk Mode

    func sendTalkRequest(text: String, module: GolfModule, completion: @escaping (Result<TalkResponse, Error>) -> Void) {
        guard let url = URL(string: APIConfig.baseURL + "/talk") else {
            completion(.failure(APIError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let payload: [String: String] = [
            "text": text,
            "module": module.uploadModuleParam,
            "device_id": EntitlementManager.shared.deviceId,
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: payload)

        session.dataTask(with: request) { data, response, error in
            self.handleDecodable(data: data, response: response, error: error, completion: completion)
        }.resume()
    }

    // MARK: - Auth

    func requestLoginCode(email: String, completion: @escaping (Result<Void, Error>) -> Void) {
        guard let url = URL(string: APIConfig.baseURL + "/auth/request-code") else {
            completion(.failure(APIError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: [
            "email": email,
            "device_id": EntitlementManager.shared.deviceId,
        ])

        session.dataTask(with: request) { _, response, error in
            if let error { completion(.failure(error)); return }
            guard let http = response as? HTTPURLResponse else {
                completion(.failure(APIError.invalidResponse)); return
            }
            if http.statusCode == 202 || http.statusCode == 200 {
                completion(.success(()))
            } else if http.statusCode == 429 {
                completion(.failure(APIError.requestFailed(statusCode: 429)))
            } else {
                completion(.failure(APIError.requestFailed(statusCode: http.statusCode)))
            }
        }.resume()
    }

    func verifyLoginCode(email: String, code: String, completion: @escaping (Result<String, Error>) -> Void) {
        guard let url = URL(string: APIConfig.baseURL + "/auth/verify-code") else {
            completion(.failure(APIError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: [
            "email": email,
            "code": code,
            "device_id": EntitlementManager.shared.deviceId,
        ])

        session.dataTask(with: request) { data, response, error in
            if let error { completion(.failure(error)); return }
            guard let http = response as? HTTPURLResponse, let data else {
                completion(.failure(APIError.invalidResponse)); return
            }
            if http.statusCode == 200 {
                if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                   let token = json["token"] as? String {
                    completion(.success(token))
                } else {
                    completion(.failure(APIError.invalidResponse))
                }
            } else {
                completion(.failure(APIError.requestFailed(statusCode: http.statusCode)))
            }
        }.resume()
    }

    func signOut(completion: @escaping (Result<Void, Error>) -> Void) {
        guard let url = URL(string: APIConfig.baseURL + "/auth/sign-out") else {
            completion(.failure(APIError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        if let token = AuthManager.shared.token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        session.dataTask(with: request) { _, response, error in
            if let error { completion(.failure(error)); return }
            guard let http = response as? HTTPURLResponse else {
                completion(.failure(APIError.invalidResponse)); return
            }
            if (200...299).contains(http.statusCode) {
                completion(.success(()))
            } else {
                completion(.failure(APIError.requestFailed(statusCode: http.statusCode)))
            }
        }.resume()
    }

    // MARK: - Payments

    func createCheckoutSession(deviceId: String, completion: @escaping (Result<CheckoutResponse, Error>) -> Void) {
        guard let url = URL(string: APIConfig.baseURL + "/payments/checkout-session") else {
            completion(.failure(APIError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["device_id": deviceId])

        session.dataTask(with: request) { data, response, error in
            self.handleDecodable(data: data, response: response, error: error, completion: completion)
        }.resume()
    }

    func checkUnlockStatus(deviceId: String, completion: @escaping (Result<UnlockStatusResponse, Error>) -> Void) {
        guard var components = URLComponents(string: APIConfig.baseURL + "/payments/unlock-status") else {
            completion(.failure(APIError.invalidURL))
            return
        }
        components.queryItems = [URLQueryItem(name: "device_id", value: deviceId)]

        guard let url = components.url else {
            completion(.failure(APIError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        if let token = AuthManager.shared.token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        session.dataTask(with: request) { data, response, error in
            self.handleDecodable(data: data, response: response, error: error, completion: completion)
        }.resume()
    }

    func checkEntitlement(deviceId: String, module: GolfModule, completion: @escaping (Result<EntitlementResponse, Error>) -> Void) {
        guard var components = URLComponents(string: APIConfig.baseURL + "/entitlement") else {
            completion(.failure(APIError.invalidURL))
            return
        }
        components.queryItems = [
            URLQueryItem(name: "device_id", value: deviceId),
            URLQueryItem(name: "module", value: module.uploadModuleParam),
        ]

        guard let url = components.url else {
            completion(.failure(APIError.invalidURL))
            return
        }

        session.dataTask(with: URLRequest(url: url)) { data, response, error in
            self.handleDecodable(data: data, response: response, error: error, completion: completion)
        }.resume()
    }

    private func handleDecodable<T: Decodable>(data: Data?, response: URLResponse?, error: Error?, completion: @escaping (Result<T, Error>) -> Void) {
        if let urlError = error as? URLError,
           [.notConnectedToInternet, .cannotConnectToHost, .timedOut].contains(urlError.code) {
            completion(.failure(APIError.serverUnreachable))
            return
        }

        if let error {
            completion(.failure(error))
            return
        }

        guard let http = response as? HTTPURLResponse else {
            completion(.failure(APIError.invalidResponse))
            return
        }

        guard (200...299).contains(http.statusCode), let data else {
            completion(.failure(APIError.requestFailed(statusCode: http.statusCode)))
            return
        }

        do {
            let decoder = JSONDecoder()
            decoder.keyDecodingStrategy = .convertFromSnakeCase
            let result = try decoder.decode(T.self, from: data)
            completion(.success(result))
        } catch {
            completion(.failure(APIError.invalidResponse))
        }
    }

    // MARK: - Analytics

    func trackEvent(_ name: String, module: GolfModule? = nil) {
        guard let url = URL(string: APIConfig.baseURL + "/analytics/event") else { return }

        var payload: [String: String] = [
            "device_id": EntitlementManager.shared.deviceId,
            "event_name": name,
            "platform": "ios",
        ]
        if let module { payload["module"] = module.uploadModuleParam }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: payload)

        session.dataTask(with: request) { _, _, _ in }.resume()
    }

    // MARK: - Performance

    func fetchHistory(module: GolfModule, completion: @escaping (Result<PerformanceHistory, Error>) -> Void) {
        guard var components = URLComponents(string: APIConfig.baseURL + "/performance/history") else {
            completion(.failure(APIError.invalidURL))
            return
        }
        components.queryItems = [
            URLQueryItem(name: "device_id", value: EntitlementManager.shared.deviceId),
            URLQueryItem(name: "module", value: module.uploadModuleParam),
            URLQueryItem(name: "limit", value: "20"),
        ]
        guard let url = components.url else {
            completion(.failure(APIError.invalidURL))
            return
        }
        session.dataTask(with: URLRequest(url: url)) { data, response, error in
            self.handleDecodable(data: data, response: response, error: error, completion: completion)
        }.resume()
    }

    private func mimeType(for url: URL) -> String {
        switch url.pathExtension.lowercased() {
        case "mov": return "video/quicktime"
        case "mp4", "m4v": return "video/mp4"
        case "avi": return "video/x-msvideo"
        default: return "application/octet-stream"
        }
    }
}

private extension Data {
    mutating func append(_ string: String) {
        if let data = string.data(using: .utf8) {
            append(data)
        }
    }
}

File: android/app/src/main/java/com/golfcoachnow/app/data/api/ApiClient.kt
package com.golfcoachnow.app.data.api

import com.golfcoachnow.app.BuildConfig
import com.golfcoachnow.app.data.model.*
import com.golfcoachnow.app.util.AuthManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.io.IOException
import java.util.concurrent.TimeUnit

object ApiClient {

    private val baseUrl = BuildConfig.API_BASE_URL

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .writeTimeout(120, TimeUnit.SECONDS)
        .build()

    private val json = Json { ignoreUnknownKeys = true }

    suspend fun sendModuleData(
        scores: Map<String, Double>,
        module: GolfModule,
        deviceId: String,
    ): Result<CorrectionResponse> = withContext(Dispatchers.IO) {
        try {
            val payload = buildString {
                append("{\"data\":{")
                append(scores.entries.joinToString(",") { "\"${it.key}\":${it.value}" })
                append("},\"device_id\":\"$deviceId\"}")
            }

            val request = Request.Builder()
                .url("$baseUrl${module.apiEndpoint}")
                .post(payload.toRequestBody("application/json".toMediaType()))
                .build()

            val response = client.newCall(request).execute()
            handleResponse<CorrectionResponse>(response)
        } catch (e: IOException) {
            Result.failure(e)
        }
    }

    suspend fun uploadVideo(
        file: File,
        module: GolfModule,
        deviceId: String,
    ): Result<CorrectionResponse> = withContext(Dispatchers.IO) {
        try {
            val body = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart(
                    "file",
                    file.name,
                    file.asRequestBody("video/mp4".toMediaType())
                )
                .build()

            val request = Request.Builder()
                .url("$baseUrl/upload?module=${module.uploadParam}&device_id=$deviceId")
                .post(body)
                .apply {
                    AuthManager.token?.let { addHeader("Authorization", "Bearer $it") }
                }
                .build()

            val response = client.newCall(request).execute()
            handleResponse<CorrectionResponse>(response)
        } catch (e: IOException) {
            Result.failure(e)
        }
    }

    suspend fun sendTalkRequest(
        text: String,
        module: GolfModule,
        deviceId: String,
    ): Result<TalkResponse> = withContext(Dispatchers.IO) {
        try {
            val payload = """{"text":"$text","module":"${module.uploadParam}","device_id":"$deviceId"}"""

            val request = Request.Builder()
                .url("$baseUrl/talk")
                .post(payload.toRequestBody("application/json".toMediaType()))
                .build()

            val response = client.newCall(request).execute()
            handleResponse<TalkResponse>(response)
        } catch (e: IOException) {
            Result.failure(e)
        }
    }

    suspend fun createCheckoutSession(deviceId: String): Result<CheckoutResponse> =
        withContext(Dispatchers.IO) {
            try {
                val payload = """{"device_id":"$deviceId"}"""

                val request = Request.Builder()
                    .url("$baseUrl/payments/checkout-session")
                    .post(payload.toRequestBody("application/json".toMediaType()))
                    .build()

                val response = client.newCall(request).execute()
                handleResponse<CheckoutResponse>(response)
            } catch (e: IOException) {
                Result.failure(e)
            }
        }

    suspend fun checkUnlockStatus(deviceId: String): Result<UnlockStatusResponse> =
        withContext(Dispatchers.IO) {
            try {
                val request = Request.Builder()
                    .url("$baseUrl/payments/unlock-status?device_id=$deviceId")
                    .get()
                    .build()

                val response = client.newCall(request).execute()
                handleResponse<UnlockStatusResponse>(response)
            } catch (e: IOException) {
                Result.failure(e)
            }
        }

    suspend fun getInstructionalVideo(module: GolfModule): Result<InstructionalVideoResponse> =
        withContext(Dispatchers.IO) {
            try {
                val request = Request.Builder()
                    .url("$baseUrl/videos/instructional?module=${module.uploadParam}")
                    .get()
                    .build()

                val response = client.newCall(request).execute()
                handleResponse<InstructionalVideoResponse>(response)
            } catch (e: IOException) {
                Result.failure(e)
            }
        }

    suspend fun shareWithFriend(email: String, deviceId: String): Result<AcceptedResponse> =
        withContext(Dispatchers.IO) {
            try {
                val payload = json.encodeToString(ShareInviteRequest(email, deviceId))
                val request = Request.Builder()
                    .url("$baseUrl/share/invite")
                    .post(payload.toRequestBody("application/json".toMediaType()))
                    .build()
                handleResponse<AcceptedResponse>(client.newCall(request).execute())
            } catch (e: IOException) {
                Result.failure(e)
            }
        }

    suspend fun messageFounder(
        message: String,
        email: String?,
        deviceId: String,
    ): Result<AcceptedResponse> = withContext(Dispatchers.IO) {
        try {
            val payload = json.encodeToString(
                FounderMessageRequest(message, email?.takeIf { it.isNotBlank() }, deviceId)
            )
            val request = Request.Builder()
                .url("$baseUrl/connect/founder")
                .post(payload.toRequestBody("application/json".toMediaType()))
                .build()
            handleResponse<AcceptedResponse>(client.newCall(request).execute())
        } catch (e: IOException) {
            Result.failure(e)
        }
    }

    suspend fun requestCode(email: String): Result<RequestCodeResponse> =
        withContext(Dispatchers.IO) {
            try {
                val payload = """{"email":"$email"}"""
                val request = Request.Builder()
                    .url("$baseUrl/auth/request-code")
                    .post(payload.toRequestBody("application/json".toMediaType()))
                    .build()
                val response = client.newCall(request).execute()
                handleResponse<RequestCodeResponse>(response)
            } catch (e: IOException) {
                Result.failure(e)
            }
        }

    suspend fun verifyCode(
        email: String,
        code: String,
        deviceId: String,
        name: String? = null,
    ): Result<VerifyCodeResponse> = withContext(Dispatchers.IO) {
        try {
            val payload = json.encodeToString(
                VerifyCodeRequest(email, code, deviceId, name?.takeIf { it.isNotBlank() })
            )
            val request = Request.Builder()
                .url("$baseUrl/auth/verify-code")
                .post(payload.toRequestBody("application/json".toMediaType()))
                .build()
            val response = client.newCall(request).execute()
            handleResponse<VerifyCodeResponse>(response)
        } catch (e: IOException) {
            Result.failure(e)
        }
    }

    suspend fun setName(name: String): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val payload = json.encodeToString(SetNameRequest(name))
            val request = Request.Builder()
                .url("$baseUrl/auth/name")
                .post(payload.toRequestBody("application/json".toMediaType()))
                .apply {
                    AuthManager.token?.let { addHeader("Authorization", "Bearer $it") }
                }
                .build()
            val response = client.newCall(request).execute()
            response.use {
                if (it.isSuccessful) Result.success(Unit)
                else Result.failure(ApiException(it.code, "Could not save that name"))
            }
        } catch (e: IOException) {
            Result.failure(e)
        }
    }

    suspend fun signOut(token: String): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder()
                .url("$baseUrl/auth/sign-out")
                .post("".toRequestBody())
                .addHeader("Authorization", "Bearer $token")
                .build()
            client.newCall(request).execute().close()
            Result.success(Unit)
        } catch (e: IOException) {
            Result.failure(e)
        }
    }

    fun trackEvent(name: String, module: GolfModule? = null) {
        val payload = buildString {
            append("{\"device_id\":\"${com.golfcoachnow.app.util.EntitlementManager.deviceId}\"")
            append(",\"event_name\":\"$name\"")
            append(",\"platform\":\"android\"")
            if (module != null) append(",\"module\":\"${module.uploadParam}\"")
            append("}")
        }
        try {
            val request = Request.Builder()
                .url("$baseUrl/analytics/event")
                .post(payload.toRequestBody("application/json".toMediaType()))
                .build()
            client.newCall(request).enqueue(object : okhttp3.Callback {
                override fun onFailure(call: okhttp3.Call, e: IOException) {}
                override fun onResponse(call: okhttp3.Call, response: okhttp3.Response) {
                    response.close()
                }
            })
        } catch (_: Exception) {}
    }

    private inline fun <reified T> handleResponse(response: okhttp3.Response): Result<T> {
        val body = response.body?.string() ?: return Result.failure(IOException("Empty response"))

        if (!response.isSuccessful) {
            return Result.failure(ApiException(response.code, body))
        }

        return try {
            Result.success(json.decodeFromString<T>(body))
        } catch (e: Exception) {
            Result.failure(IOException("Failed to parse response"))
        }
    }
}

class ApiException(val code: Int, override val message: String) : IOException(message)
